wgpu-hal/src/vulkan/device.rs

```diff
@@ -1452,119 +1419,155 @@ impl crate::Device for super::Device {
 
         let set = vk_sets.pop().unwrap();
         if let Some(label) = desc.label {
-            unsafe {
-                self.shared
-                    .set_object_name(vk::ObjectType::DESCRIPTOR_SET, *set.raw(), label)
-            };
+            unsafe { self.shared.set_object_name(*set.raw(), label) };
+        }
+
+        struct ExtendStack<'a, T> {
+            remainder: &'a mut [MaybeUninit<T>],
         }
 
+        impl<'a, T> ExtendStack<'a, T> {
+            fn from_vec_capacity(vec: &'a mut Vec<T>) -> Self {
+                Self {
+                    remainder: vec.spare_capacity_mut(),
+                }
+            }
+
+            fn extend_one(self, value: T) -> (Self, &'a mut T) {
+                let (to_init, remainder) = self.remainder.split_first_mut().unwrap();
+                let init = to_init.write(value);
+                (Self { remainder }, init)
+            }
+
+            fn extend(
+                self,
+                iter: impl IntoIterator<Item = T> + ExactSizeIterator,
+            ) -> (Self, &'a mut [T]) {
+                let (to_init, remainder) = self.remainder.split_at_mut(iter.len());
+
+                for (value, to_init) in iter.into_iter().zip(to_init.iter_mut()) {
+                    to_init.write(value);
+                }
+
+                // we can't use the safe (yet unstable) MaybeUninit::write_slice() here because of having an iterator to write
+
+                let init = {
+                    #[allow(trivial_casts)]
+                    // SAFETY: The loop above has initialized exactly as many items as to_init is
+                    // long, so it is safe to cast away the MaybeUninit<T> wrapper into T.
+
+                    // Additional safety docs from unstable slice_assume_init_mut
+                    // SAFETY: similar to safety notes for `slice_get_ref`, but we have a
+                    // mutable reference which is also guaranteed to be valid for writes.
+                    unsafe {
+                        &mut *(to_init as *mut [MaybeUninit<T>] as *mut [T])
+                    }
+                };
+                (Self { remainder }, init)
+            }
+        }
+
+        // TODO: Pin?
```

Comment:
Do we need to pin something?

Reply:
If my understanding of `Pin` is correct (see above: **it is not**) it _should_ prevent us from causing destructive actions on the pointer (`.push()`, `.resize()`, `.clear()`) after having `let mut` access to a `Vec`, while we can still deref it into a `&mut [T]` on which we cannot perform

I don't think that deref mechanism works for `spare_capacity_mut()` though.

Please correct me about this.

Reply:
As long as the mutable references to the subslices of the Vec are still around we can't call any of `Vec`'s methods that take `&mut Self`, so we should be good.